src/hooks/useSortRedirects.ts

```diff
@@ -0,0 +1,53 @@
+import useNavigator from "@saleor/hooks/useNavigator";
+import { Sort } from "@saleor/types";
+import { useEffect } from "react";
+
+export type SortByRankUrlQueryParams<T extends string> = Sort<T> & {
+  query?: string;
+};
+
+export interface UseSortRedirectsOpts<SortField extends string> {
+  params: SortByRankUrlQueryParams<SortField>;
+  defaultSortField: SortField;
+  urlFunc: (params: SortByRankUrlQueryParams<SortField>) => string;
+  resetToDefault?: boolean;
+}
+/**
+ * useSortRedirects is a hook that should be used in lists views
+ * where using search changes the sort field to "rank". Removing
+ * query changes sort field back to default one provided in params.
+ */
+export function useSortRedirects<SortField extends string>({
+  params,
+  defaultSortField,
+  urlFunc,
+  resetToDefault,
+}: UseSortRedirectsOpts<SortField>) {
+  const navigate = useNavigator();
+
+  const hasQuery = !!params.query?.trim();
+
+  useEffect(() => {
+    const sortWithQuery = "rank" as SortField;
+    const sortWithoutQuery =
+      params.sort === "rank" ? defaultSortField : params.sort;
+    navigate(
+      urlFunc({
+        ...params,
+        asc: hasQuery ? false : params.asc,
+        sort: hasQuery ? sortWithQuery : sortWithoutQuery,
+      }),
+    );
+  }, [params.query]);
```

Comment:
Shouldn't this also include `params.asc` and `params.sort`?

```suggestion
  }, [params.query, params.sort, params.asc]);
```

Reply:
It's not required because sorting by other fields is managed by other parts of code. Also, sorting by "rank" field with ascending order makes no sense - why would anyone need worst matches first